readme.md
### three_stage
three.js 基本场景搭建。

npm 地址: https://www.npmjs.com/package/three_stage

demo:
```js
import Stage from 'three_stage'
import * as THREE from 'three'
import * as dat from "dat.gui";

class Options {
  constructor() {
    this.speed = 0.01;
  }
};

class App {
  constructor() {
    this.settings = this.settings.bind(this)
    this.stage = new Stage("#app")
    this.stage.run()
    this.addBox(10)

    this.settings()
  }


  settings() {
    var options = new Options();
    var gui = new dat.GUI();

    gui.add(options, 'speed', 0, 0.1);
    this.options = options
  }

  addBox(size) {
    var geometry = new THREE.BoxBufferGeometry(size, size, size);
    var material = new THREE.MeshPhongMaterial({
      color: 0x63e42a,
      emissive: 0x072534,
      side: THREE.DoubleSide,
      shading: THREE.FlatShading
    })
    var cube = new THREE.Mesh(geometry, material);
    cube.name = "test_cube"
    this.stage.scene.add(cube)
    this.stage.onUpdate(() => {
      cube.rotation.x += this.options.speed || 0
      cube.rotation.y += this.options.speed || 0
    })
  }
}

window.onload = () => {
  let app = new App()
}
```


npm config set registry http://registry.npm.taobao.org
npm config set registry https://registry.npmjs.org;
webpack-dev-server --config webpack.dev.js --inline --host 192.168.100.134 --progress


changeLog.md
#### 1.0.5
使用自己封装的相机控制器

#### 1.0.4
添加基本场景


index.js
import * as THREE from 'three'
// import { OrbitControls } from 'three/examples/jsm/controls/OrbitControls'
import OrbitControls from './camera-control/MyOrbitControls'
import Stats from 'three/examples/jsm/libs/stats.module.js'

export default class Stage {
  constructor(container, param) {
    this.fuArr = []
    this.viceCamera = null
    this.tmpTarget = new THREE.Vector3()
    this.orbitControls = null
    this.cameraHelper = null
    this.container = container;
    this.initFlag = false;
    // 场景
    this.scene = new THREE.Scene();
    this.scene.name = "moumade";
    window.scene = this.scene;

    // 环境光
    var ambient = new THREE.AmbientLight(0xffffff, 0.5);
    ambient.name = "ambient";
    this.scene.add(ambient);

    let dirLight = new THREE.DirectionalLight(0xffffff, 0.5);
    dirLight.position.set(100, 100, 100);
    dirLight.name = "dirLight";
    this.scene.add(dirLight);

    // 渲染器
    this.containerEle = document.querySelector(container);
    let vW = this.containerEle.clientWidth;
    let vH = this.containerEle.clientHeight;
    this.renderer = new THREE.WebGLRenderer({
      antialias: true,
      // alpha: false,
      // preserveDrawingBuffer: true,
      // failIfMajorPerformanceCaveat: true,
    });
    this.renderer.setClearColor(0x33334c, 1.0);
    // this.renderer.autoClear = true;
    this.renderer.setSize(vW, vH, false);
    this.renderer.setPixelRatio(window.devicePixelRatio);
    // console.error(vW, vH);

    this.containerEle.appendChild(this.renderer.domElement);
    this.camera = new THREE.PerspectiveCamera(45, 1, 0.01, 1000)

    this.camera.lookAt(0, 0, 0)
    this.camera.name = "camera";
    this.scene.add(this.camera);
    this.handleResize = this.handleResize.bind(this)
    this._loop = this._loop.bind(this)
    window.addEventListener("resize", this.handleResize);
    this.handleResize();
    if (param && param.disableControl) {

    } else {
      this.initControls();
    }
    this.camera.position.set(0, 0, 50)

    let axis = new THREE.AxesHelper(10000);
    this.scene.add(axis);

    this.stats = new Stats();
    this.stats.domElement.style.position = 'absolute';
    this.stats.domElement.style.left = '0px';
    this.stats.domElement.style.top = '0px';
    this.stats.domElement.style.width = '100px';
    this.stats.domElement.id = 'stats';
    this.containerEle.appendChild(this.stats.dom);

    this.state = true
  }

  initControls() {
    let control = new OrbitControls(this.camera, this.renderer.domElement);
    this.control = control
    // 使动画循环使用时阻尼或自转 意思是否有惯性
    control.enableDamping = true;
    //动态阻尼系数 就是鼠标拖拽旋转灵敏度
    control.dampingFactor = 0.35;
    //是否可以缩放
    control.enableZoom = true;
    control.zoomSpeed = 0.35;
    //是否自动旋转
    control.autoRotate = false;
    //设置相机距离原点的最远距离
    // control.minDistance = 22; //1000
    //设置相机距离原点的最远距离
    // control.maxDistance = 50; //3000
    //是否开启右键拖拽
    // control.enablePan = false;
  }

  handleResize() {
    // 获取新的大小
    let vpW = this.containerEle.clientWidth;
    let vpH = this.containerEle.clientHeight;
    // 设置场景
    this.renderer.domElement.width = vpW;
    this.renderer.domElement.height = vpH;
    this.renderer.setSize(this.containerEle.clientWidth, this.containerEle.clientHeight);
    // 设置相机
    this.camera.aspect = vpW / vpH;
    this.camera.updateProjectionMatrix();
  }


  run() {
    this._loop()
  }

  addViceCamera(viceCamera) {
    this.viceCamera = viceCamera
  }

  setRender(v) {
    this.state = v
  }
  onUpdate(fu) {
    this.fuArr.push(fu)
  }

  _loop() {
    this.stats.update()
    this.camera.updateProjectionMatrix();
    this.camera.updateMatrixWorld()
    this.state && this.renderer.render(this.scene, this.camera);

    this.fuArr.forEach(fun => {
      fun()
    });
    this.control && this.control.update()
    requestAnimationFrame(this._loop)
  }
}


camera-control/Pan.js
import * as THREE from 'three'

// 平移
export default class Pan {
  constructor(param) {
    const { camera, dampingFactor, target, domWidth, domHeight, enableDamping } = param
    this.param = param
    this.panOffset = new THREE.Vector3();
    this.panStart = new THREE.Vector2();
    this.panEnd = new THREE.Vector2();
    this.panDelta = new THREE.Vector2();
    this.panSpeed = 1
  }

  setPanStart(x, y) {
    this.panStart.set(x, y)
  }

  setPanEnd(x, y) {
    this.panEnd.set(x, y)
    this.panDelta.subVectors(this.panEnd, this.panStart).multiplyScalar(this.panSpeed)
    this.setPanStart(x, y)
    this.pan(this.panDelta.x, this.panDelta.y);
    this.update()
  }

  panLeft(distance, objectMatrix) {
    var v = new THREE.Vector3();
    v.setFromMatrixColumn(objectMatrix, 0); // get X column of objectMatrix
    v.multiplyScalar(-distance);
    this.panOffset.add(v);
  }

  panUp(distance, objectMatrix) {
    var v = new THREE.Vector3();
    v.setFromMatrixColumn(objectMatrix, 1);
    v.multiplyScalar(distance);
    this.panOffset.add(v);
  }

  pan(deltaX, deltaY) {
    const offset = new THREE.Vector3()
    const { camera, target, domWidth, domHeight } = this.param
    var position = camera.position;
    offset.copy(position).sub(target);
    var targetDistance = offset.length();

    // half of the fov is center to top of screen
    targetDistance *= Math.tan((camera.fov / 2) * Math.PI / 180.0);

    // we use only clientHeight here so aspect ratio does not distort speed
    this.panLeft(2 * deltaX * targetDistance / domWidth, camera.matrix);
    this.panUp(2 * deltaY * targetDistance / domHeight, camera.matrix);
  }

  update() {
    const { camera, target } = this.param
    camera.position.add(this.panOffset)
    target.add(this.panOffset);
    this.panOffset.set(0, 0, 0)
  }
}


camera-control/Dolly.js
import * as THREE from 'three'

/**
 * 缩放控制
 */
export class Dolly {
  constructor(param) {
    this.param = param
    this.zoomSpeed = 1
    this.scale = 1

    this.dollyStart = new THREE.Vector2();
    this.dollyEnd = new THREE.Vector2();
  }

   setDollyStartForMobile(x, y) {
     this.dollyStart.set(x, y)
   }

  setDollyEndForMobile(x, y) {
    const dollyDelta = new THREE.Vector2();
     this.dollyEnd.set(x, y)
    dollyDelta.set(0, Math.pow(this.dollyEnd.y / this.dollyStart.y, this.zoomSpeed));
    this.dollyOut(dollyDelta.y);
    this.dollyStart.set(x,y)
   }

  handlePC(deltaY) {
    if (deltaY > 0) {
      this.dollyOut(this.getZoomScale())
    } else {
      this.dollyIn(this.getZoomScale())
    }
    this.update()
  }

  dollyOut(dollyScale) {
    this.scale /= dollyScale
  }
  dollyIn(dollyScale) {
    this.scale *= dollyScale
  }

  getZoomScale() {
    return Math.pow(0.95, this.zoomSpeed);
  }

  update() {
    let { target } = this.param

    let spherical = new THREE.Spherical();
    const offset = new THREE.Vector3();
    var position = this.param.camera.position
    offset.copy(position).sub(target);
    spherical.setFromVector3(offset);
    spherical.radius *= this.scale
    this.scale = 1

    offset.setFromSpherical(spherical);
    position.copy(target).add(offset);
  }
}


camera-control/Rotation.js
import * as THREE from 'three'

/**
 * 鼠标左键, 旋转控制
 */

// 水平旋转的计算方式,
export const RotateLeftType = {
  OFFSET: 0,
  ANGLE: 1,
}

export class Rotation {
  constructor(param) {
    const { camera, dampingFactor, target, domWidth, domHeight, enableDamping } = param
    this.param = param

    this.spherical = new THREE.Spherical();
    this.sphericalDelta = new THREE.Spherical();

    this.rotateUpStart = new THREE.Vector2();
    this.rotateUpEnd = new THREE.Vector2();

    this.rotateLeftStart = new THREE.Vector2();
    this.rotateLeftEnd = new THREE.Vector2();

    this.rotateSpeed = 1.0
    this.quat = new THREE.Quaternion().setFromUnitVectors(this.param.camera.up, new THREE.Vector3(0, 1, 0));
    this.quatInverse = this.quat.clone().invert();

    this.rotateLeftType = RotateLeftType.ANGLE
  }

  /**
   * 设置水平选择的方式,角度或者offset
   * @param {*} v
   */
  setRotateLeftType(v) {
    this.rotateLeftType = v
   }

  /**
   * 设置水平旋转的开始点
   * @param {*} x
   * @param {*} y
   */
  setRotateLeftStart(x, y) {
    this.rotateLeftStart.set(x, y)
  }

  /**
   * 设置水平旋转的结束点
   * @param {*} x
   * @param {*} y
   */
  setRotateLeftEnd(x, y) {
    this.rotateLeftEnd.set(x, y)

    if (this.rotateLeftType == RotateLeftType.ANGLE) {
      // 计算相对于正 x 轴的角度（以弧度为单位）
      let anglestart = this.rotateLeftStart.angle();
      let angleend = this.rotateLeftEnd.angle();
      this.rotateLeft((-angleend + anglestart) * this.rotateSpeed);
    } else if (this.rotateLeftType == RotateLeftType.OFFSET) {
      const rotateDelta = new THREE.Vector2()
      rotateDelta.subVectors(this.rotateLeftEnd, this.rotateLeftStart).multiplyScalar(this.rotateSpeed);
      this.rotateLeft(2 * Math.PI * rotateDelta.x / this.param.domWidth);
    }
     this.rotateLeftStart.set(x, y)
  }

  /**
   * 设置垂直旋转的开始点
   * @param {*} x
   * @param {*} y
   */
  setRotateUpStart(x, y) {
    this.rotateUpStart.set(x, y)
  }

  /**
   * 设置垂直旋转的结束点
   * @param {*} x
   * @param {*} y
   */
  setRotateUpEnd(x, y) {
    this.rotateUpEnd.set(x, y)
    const rotateDelta = new THREE.Vector3()
    rotateDelta.subVectors(this.rotateUpEnd, this.rotateUpStart).multiplyScalar(this.rotateSpeed);

    this.rotateUp(2 * Math.PI * rotateDelta.y / this.param.domHeight);
    this.rotateUpStart.copy(this.rotateUpEnd);
    this.update()
  }

  /**
   * 水平旋转
   * @param {*} angle
   */
  rotateLeft(angle) {
    this.sphericalDelta.theta -= angle;
  }

  /**
   * 垂直旋转
   * @param {*} angle
   */
  rotateUp(angle) {
    this.sphericalDelta.phi -= angle;
  }

  update() {
    const minPolarAngle = 0; // radians
    const maxPolarAngle = Math.PI; // radians

    let { target, dampingFactor, camera, enableDamping } = this.param
    enableDamping = false
    const offset = new THREE.Vector3();
    var position = this.param.camera.position

    offset.copy(position).sub(target);
    offset.applyQuaternion(this.quat);
    this.spherical.setFromVector3(offset);

    if (enableDamping) {
      this.spherical.theta += this.sphericalDelta.theta * this.dampingFactor;
      this.spherical.phi += this.sphericalDelta.phi * dampingFactor;
    } else {
      this.spherical.theta += this.sphericalDelta.theta;
      this.spherical.phi += this.sphericalDelta.phi;
    }

    this.spherical.phi = Math.max(minPolarAngle, Math.min(maxPolarAngle, this.spherical.phi));
    this.spherical.makeSafe();

    offset.setFromSpherical(this.spherical);
    offset.applyQuaternion(this.quatInverse);

    position.copy(target).add(offset);
    camera.lookAt(target);

    if (enableDamping) {
      this.sphericalDelta.theta *= (1 - dampingFactor);
      this.sphericalDelta.phi *= (1 - dampingFactor);
    } else {
      this.sphericalDelta.set(0, 0, 0);
    }

  }
}


camera-control/MyOrbitControls.js
import * as THREE from 'three'
import { Rotation, RotateLeftType } from './Rotation.js'
import Pan from './Pan.js'
import { Dolly } from './Dolly.js'
import { MOUSE, TOUCH, STATE } from './constants'

export default class MyOrbitControls {
  constructor(camera, domElement) {
    // 全局参数
    this.camera = camera
    this.domElement = domElement
    this.target = new THREE.Vector3(0, 0, 0)

    // 事件对应的处理动作
    this.mouseButtons = {
      LEFT: MOUSE.ROTATE,
      MIDDLE: MOUSE.DOLLY,
      RIGHT: MOUSE.PAN
    };
    this.touches = {
      ONE: TOUCH.ROTATE, // 一个手指只能 PAN 或者 ROTATE
      TWO: TOUCH.DOLLY_ROTATE // 两个手指多了缩放
    };
    this.pan = new Pan({
      camera: this.camera,
      target: this.target,
      domWidth: this.domElement.width,
      domHeight: this.domElement.height,
    })

    this.dolly = new Dolly({
      camera: this.camera,
      target: this.target,
      domWidth: this.domElement.width,
      domHeight: this.domElement.height,
    })

    this.rotation = new Rotation({
      camera: this.camera,
      enableDamping: false,
      dampingFactor: 0.05,
      target: this.target,
      domWidth: this.domElement.width,
      domHeight: this.domElement.height,
    })

    // 鼠标参数
    this.state = STATE.NONE;

    this.onPointerDown = this.onPointerDown.bind(this)
    this.onPointerMove = this.onPointerMove.bind(this)
    this.onPointerUp = this.onPointerUp.bind(this)
    this.onWheel = this.onWheel.bind(this)

    this.onTouchStart = this.onTouchStart.bind(this)
    this.onTouchEnd = this.onTouchEnd.bind(this)
    this.onTouchMove = this.onTouchMove.bind(this)
    this.initEvent()
  }


  initEvent() {
    // pc
    this.domElement.addEventListener('pointerdown', this.onPointerDown);
    this.domElement.ownerDocument.addEventListener('pointerup', this.onPointerUp);
    this.domElement.ownerDocument.addEventListener('pointermove', this.onPointerMove);
    this.domElement.addEventListener('contextmenu', this.onContextMenu);
    this.domElement.addEventListener('wheel', this.onWheel);

    // mobile
    this.domElement.addEventListener('touchstart', this.onTouchStart);
    this.domElement.addEventListener('touchend', this.onTouchEnd);
    this.domElement.addEventListener('touchmove', this.onTouchMove);
  }

  dispose() {
    this.domElement.removeEventListener('pointerdown', this.onPointerDown);
    this.domElement.ownerDocument.removeEventListener('pointerup', this.onPointerUp);
    this.domElement.ownerDocument.removeEventListener('pointermove', this.onPointerMove);
    this.domElement.removeEventListener('contextmenu', this.onContextMenu);
    this.domElement.removeEventListener('wheel', this.onWheel);

     this.domElement.removeEventListener('touchstart', this.onTouchStart);
     this.domElement.removeEventListener('touchend', this.onTouchEnd);
     this.domElement.removeEventListener('touchmove', this.onTouchMove);
  }

  onWheel(event) {
    const { deltaY } = event
    this.dolly.handlePC(deltaY)
  }

  onContextMenu(event) {
    event.preventDefault();
  }

  onPointerMove(event) {
    if (this.state == STATE.NONE) return
    const mouse = { x: event.clientX, y: event.clientY }

    switch (this.state) {
      case MOUSE.ROTATE:
        this.rotation.setRotateUpEnd(mouse.x, mouse.y)
        this.rotation.setRotateLeftEnd(mouse.x, mouse.y)
        break;

      case MOUSE.PAN:
        this.pan.setPanEnd(mouse.x, mouse.y)
        break
    }
  }

  onPointerUp() {
    this.state = STATE.NONE
  }

  onPointerDown(event) {
    // 这里不处理touch事件
    const allowEvent = ["mouse", "pen"]
    if (!allowEvent.includes(event.pointerType)) {
      return
    }

    event.preventDefault();
    let mouseAction = STATE.NONE
    switch (event.button) {
      case 0:
        mouseAction = this.mouseButtons.LEFT
        break;

      case 1:
        mouseAction = this.mouseButtons.MIDDLE;
      break;

      case 2:
        mouseAction = this.mouseButtons.RIGHT
        break;

      default:
        mouseAction = STATE.NONE;
    }

    const mouse = { x: event.clientX, y: event.clientY }
    switch (mouseAction) {
      case MOUSE.ROTATE:
        this.rotation.setRotateLeftType(RotateLeftType.OFFSET)
        this.rotation.setRotateUpStart(mouse.x, mouse.y)
        this.rotation.setRotateLeftStart(mouse.x, mouse.y)
        this.state = STATE.ROTATE
        break;

      case MOUSE.PAN:
        this.pan.setPanStart(mouse.x, mouse.y)
        this.state = STATE.PAN
        break;

      default:
        this.state = STATE.NONE
        break;
    }
  }

  handleTouchStartRotate(event) {
    if (event.touches.length == 1) {
      const x = event.touches[0].pageX
      const y = event.touches[0].pageY
      this.rotation.setRotateUpStart(x, y);
      this.rotation.setRotateLeftStart(x, y);
    } else {
      var x = 0.5 * (event.touches[0].pageX + event.touches[1].pageX);
      var y = 0.5 * (event.touches[0].pageY + event.touches[1].pageY);
      this.rotation.setRotateUpStart(x, y);

      var dlx = event.touches[0].pageX - event.touches[1].pageX;
      var dly = event.touches[0].pageY - event.touches[1].pageY;
      this.rotation.setRotateLeftStart(dlx, dly);
    }
  }

  handleTouchStartPan(event) {
    if (event.touches.length == 1) {
     this.pan.setPanStart(event.touches[0].pageX, event.touches[0].pageY);
    } else {
      var x = 0.5 * (event.touches[0].pageX + event.touches[1].pageX);
      var y = 0.5 * (event.touches[0].pageY + event.touches[1].pageY);
     this.pan.setPanStart(x, y);
    }
  }

  handleTouchStartDolly(event) {
    	var dx = event.touches[0].pageX - event.touches[1].pageX;
    	var dy = event.touches[0].pageY - event.touches[1].pageY;
    var distance = Math.sqrt(dx * dx + dy * dy);
    	this.dolly.setDollyStartForMobile(0, distance);
  }

  handleTouchMoveDolly(event) {
    var dx = event.touches[0].pageX - event.touches[1].pageX;
    var dy = event.touches[0].pageY - event.touches[1].pageY;
    var distance = Math.sqrt(dx * dx + dy * dy);
    this.dolly.setDollyEndForMobile(0, distance);
  }

  handleTouchMoveDollyPan(event) {
    this.handleTouchMoveDolly(event);
    this.handleTouchMovePan(event);
  }

  handleTouchStartDollyPan(event) {
    this.handleTouchStartDolly(event);
    this.handleTouchStartPan(event)
  }

  handleTouchStartDollyRotate(event) {
    this.handleTouchStartDolly(event)
    this.handleTouchStartRotate(event)
  }
  handleTouchMoveDollyRotate(event) {
    this.handleTouchMoveDolly(event)
    this.handleTouchMoveRotate(event)
  }

  onTouchStart(event) {

    event.preventDefault(); // prevent scrolling
    switch (event.touches.length) {
      case 1:
        switch (this.touches.ONE) {
          case TOUCH.ROTATE:
            this.rotation.setRotateLeftType(RotateLeftType.OFFSET)
            this.handleTouchStartRotate(event);
            this.state = STATE.TOUCH_ROTATE;
            break;
          case TOUCH.PAN:
            if (this.enablePan === false) return;
            this.handleTouchStartPan(event);
            this.state = STATE.TOUCH_PAN;
            break;

          default:
            this.state = STATE.NONE;
        }
        break;

      case 2:
        switch (this.touches.TWO) {
          case TOUCH.DOLLY_PAN:
            this.handleTouchStartDollyPan(event);
            this.state = STATE.TOUCH_DOLLY_PAN;
            break;

          case TOUCH.DOLLY_ROTATE:
            this.rotation.setRotateLeftType(RotateLeftType.ANGLE)
            this.handleTouchStartDollyRotate(event);
            this.state = STATE.TOUCH_DOLLY_ROTATE;
            break;

          default:
        }
        break;

      default:
        this.state = STATE.NONE
        break
    }
  }

  onTouchEnd() {
    // no-op
  }

  handleTouchMoveRotate(event) {
    if (event.touches.length == 1) {
      const x = event.touches[0].pageX
      const y = event.touches[0].pageY
      this.rotation.setRotateUpEnd(x, y);
      this.rotation.setRotateLeftEnd(x, y);
    } else {
      var x = 0.5 * (event.touches[0].pageX + event.touches[1].pageX);
      var y = 0.5 * (event.touches[0].pageY + event.touches[1].pageY);
      this.rotation.setRotateUpEnd(x, y);

      var dlx = event.touches[0].pageX - event.touches[1].pageX;
      var dly = event.touches[0].pageY - event.touches[1].pageY;

      this.rotation.setRotateLeftEnd(dlx, dly);
    }
  }

  handleTouchMovePan(event) {
		if (event.touches.length == 1) {
		  this.pan.setPanEnd(event.touches[0].pageX, event.touches[0].pageY);
		} else {
		  var x = 0.5 * (event.touches[0].pageX + event.touches[1].pageX);
		  var y = 0.5 * (event.touches[0].pageY + event.touches[1].pageY);
		  this.pan.setPanEnd(x, y);
		}
  }

  onTouchMove(event) {
    event.preventDefault(); // prevent scrolling
    event.stopPropagation();

    	switch (this.state) {
    	  case STATE.TOUCH_ROTATE:
    	    if (this.enableRotate === false) return;
    	    this.handleTouchMoveRotate(event);
    	    this.update();
    	    break;

    	  case STATE.TOUCH_PAN:
    	    this.handleTouchMovePan(event);
    	    this.update();
    	    break;

    	  case STATE.TOUCH_DOLLY_PAN:
    	    this.handleTouchMoveDollyPan(event);
    	    this.update();
    	    break;

    	  case STATE.TOUCH_DOLLY_ROTATE:
    	    this.handleTouchMoveDollyRotate(event);
    	    this.update();
    	    break;

    	  default:
    	    this.state = STATE.NONE;
    	}
  }

  update() {
    this.pan.update()
    this.dolly.update()
    this.rotation.update()
  }
}


example/app.js
import Stage from '../index.js'
import * as THREE from 'three'
import * as dat from "dat.gui";
import VConsole from 'vconsole';
// const vConsole = new VConsole({ maxLogNumber: 1000 });
class Options {
  constructor() {
    this.message = 'dat.gui';
    this.speed = 0.01;
    this.displayOutline = false;
  }
};

class App {
  constructor() {
    this.settings = this.settings.bind(this)
    this.stage = new Stage("#app")
    this.stage.run()
    this.addBox(10)

    this.settings()
  }


  settings() {
    var options = new Options();
    var gui = new dat.GUI();

    gui.add(options, 'message');
    gui.add(options, 'speed', 0, 0.1);
    this.options = options
    window.lm = options
  }

  addBox(size) {
    var geometry = new THREE.BoxBufferGeometry(size, size, size);
    var material = new THREE.MeshPhongMaterial({
      color: 0x63e42a,
      emissive: 0x072534,
      side: THREE.DoubleSide,
      shading: THREE.FlatShading
    })
    var cube = new THREE.Mesh(geometry, material);
    cube.name = "test_cube"
    this.stage.scene.add(cube)
    this.stage.onUpdate(() => {
      // cube.rotation.x += this.options.speed || 0
      // cube.rotation.y += this.options.speed || 0
    })
  }
}

window.onload = () => {
  let app = new App()
}
